File: src/wings/routing/selector.py
"""Weighted random selection algorithm — pure function, zero dependencies."""

import random
from collections.abc import Sequence

from wings.routing.types import PoolEntry

WEIGHT_EPSILON = 1e-9


class NoAPIAvailable(Exception):
    """No API is available in the resolved pool."""


def weighted_select(entries: Sequence[PoolEntry]) -> str:
    """Select an API by weighted random choice (pure function).

    Entries with weight <= EPSILON or enabled=False are excluded.

    Args:
        entries: Pool entries to select from.

    Returns:
        The api_id of the selected entry.

    Raises:
        NoAPIAvailable: No active entries in the pool.
    """
    active = [e for e in entries if e.enabled and e.weight > WEIGHT_EPSILON]
    if not active:
        raise NoAPIAvailable("no active API in pool")

    total = sum(e.weight for e in active)
    r = random.uniform(0, total)
    cumulative = 0.0
    for e in active:
        cumulative += e.weight
        if r <= cumulative:
            return e.api_id
    # Float rounding caused a miss; return the last entry
    return active[-1].api_id

File: src/routing/selector.ts
/** Softmax-based API selection from global pool + score masks. */

import type { PoolEntry } from "./types.ts";

export const NEG_INF = -Infinity;

export class NoAPIAvailable extends Error {
  constructor(message: string = "no API available") {
    super(message);
    this.name = "NoAPIAvailable";
  }
}

/** Select an API via softmax over effective scores.
 *
 * effective_score[api] = base_score + mask_delta
 * mask_delta of -Infinity = disabled for this task type.
 *
 * Returns the selected api_id. Throws NoAPIAvailable if all APIs are disabled.
 */
export function softmaxSelect(
  entries: PoolEntry[],
  mask: Record<string, number> | null = null,
): string {
  const adjustments = mask ?? {};

  // Compute effective scores.
  const effective: Record<string, number> = {};
  for (const e of entries) {
    const delta = adjustments[e.api_id] ?? 0;
    effective[e.api_id] = e.score + delta;
  }

  // Filter out -Infinity (disabled).
  const activeEntries = (Object.entries(effective) as Array<[string, number]>)
    .filter(([, v]) => v !== NEG_INF);

  if (activeEntries.length === 0) {
    throw new NoAPIAvailable("all APIs disabled for this task type");
  }

  // Softmax with numerical stability (subtract max).
  const maxScore = Math.max(...activeEntries.map(([, v]) => v));
  const exps: Array<[string, number]> = activeEntries.map(([k, v]) => [
    k,
    Math.exp(v - maxScore),
  ]);
  const total = exps.reduce((sum, [, e]) => sum + e, 0);

  // Weighted random selection.
  const r = Math.random() * total;
  let cumulative = 0;
  for (const [apiId, expVal] of exps) {
    cumulative += expVal;
    if (r <= cumulative) return apiId;
  }
  // Float rounding fallback.
  return activeEntries[activeEntries.length - 1]![0];
}
